strawberry/decorator.py

```diff
@@ -0,0 +1,33 @@
+import functools
+
+from .resolvers import get_resolver_arguments
+from .utils.inspect import get_func_args
+
+
+def make_strawberry_decorator(func):
+    def decorator(resolver):
+        if hasattr(resolver, "_field_definition"):
+            raise Exception("Can't apply decorator after strawberry.field")  # TODO
+
+        function_args = get_func_args(resolver)
+
+        @functools.wraps(resolver)
+        def wrapped_resolver(source, info, **kwargs):
+            def wrapped(**kwargs):
+                # If resolver is another strawberry decorator then pass all the
+                # arguments to it
+                if getattr(resolver, "_strawberry_decorator", False):
+                    return resolver(source, info, **kwargs)
+
+                args, extra_kwargs = get_resolver_arguments(function_args, source, info)
+                return resolver(*args, **extra_kwargs, **kwargs)
+
+            # Call the decorator body with the original set of kwargs so that it
+            # has the opportunity to modify them
+            return func(wrapped, source, info=info, **kwargs)
+
+        wrapped_resolver._strawberry_decorator = True
```

Comment:
Could we return a Callable wrapper class instead of the function monkeypatched? A benefit now would be the ability to use `isinstance` instead of `hasattr`

Reply:
This would also play into my effort to try to remove a lot of the monkeypatching from elsewhere in the codebase

Reply:
Sure that would also work. I'll refactor the code to that.

As an aside what's the benefit of using `isinstance` instead of `hasattr`? To me they seem pretty much equivalent and I'd imagine that `isinstance` has (marginally) worse performance.

Reply:
At minimum, we can add typing annotations for mypy/IDEs to provide warnings/suggestions (without having to wait for typing.Protocol in 3.8). 

Symbols are also more easily refactored and indexed than strings (although, this specific argument could potentially be fixed with variables for `hasattr`).

Reply:
> without having to wait for typing.Protocol in 3.8

we can already use them via typing_extensions :)

Reply:
So I've just had a go at this I don't actually think it's possible. I changed it to something like this:

```python
class StrawberryDecorator:
	def __init__(self, resolver):
		self.resolver = resolver
		functools.update_wrapper(self, resolver)  # this is necessary so that `StrawberryField` can extract the arguments and name

	def __call__(self, *args, **kwargs):
		return self.resolver(*args, **kwargs)


def make_strawberry_decorator(func):
    def decorator(resolver):
        function_args = get_func_args(resolver)

        @functools.wraps(resolver)
        def wrapped_resolver(source, info, **kwargs):
            def wrapped(**kwargs):
                if isinstance(resolver, StrawberryDecorator):  # <-- this does not work
                    return resolver(source, info, **kwargs)

                args, extra_kwargs = get_resolver_arguments(function_args, source, info)
                return resolver(*args, **extra_kwargs, **kwargs)

            return func(wrapped, source, info=info, **kwargs)

        return StrawberryDecorator(wrapped_resolver)

    return decorator
```

Because of the `functools.update_wrapper` call the `isinstance(resolver, StrawberrryDecorator)` doesn't work which defeats the point of using the class in the first place.

I think the best option is to just stick with the monkeypatching even if it's not ideal.